src/hotspot/cpu/riscv/nativeInst_riscv.hpp

```diff
@@ -123,43 +135,81 @@ class NativeInstruction {
-    return compare_instr_field(instr + 4, 19, 15, instr, 11, 7)       &&  // check the rs1 field of addi and the rd field of lui
-           compare_instr_field(instr + 4, 19, 15, instr + 4, 11, 7)   &&  // check the rs1 field and the rd field of addi
-           compare_instr_field(instr + 8, 19, 15, instr + 4, 11, 7)   &&  // check the rs1 field of slli and the rd field of addi
-           compare_instr_field(instr + 8, 19, 15, instr + 8, 11, 7)   &&  // check the rs1 field and the rd field of slli
-           compare_instr_field(instr + 12, 19, 15, instr + 8, 11, 7)  &&  // check the rs1 field of addi and the rd field of slli
-           compare_instr_field(instr + 12, 19, 15, instr + 12, 11, 7) &&  // check the rs1 field and the rd field of addi
-           compare_instr_field(instr + 16, 19, 15, instr + 12, 11, 7) &&  // check the rs1 field of slli and the rd field of addi
-           compare_instr_field(instr + 16, 19, 15, instr + 16, 11, 7) &&  // check the rs1 field and the rd field fof slli
-           compare_instr_field(instr + 20, 19, 15, instr + 16, 11, 7) &&  // check the rs1 field of addi and the rd field of slli
-           compare_instr_field(instr + 20, 19, 15, instr + 20, 11, 7) &&  // check the rs1 field and the rd field of addi
-           compare_instr_field(instr + 24, 19, 15, instr + 20, 11, 7) &&  // check the rs1 field of slli and the rd field of addi
-           compare_instr_field(instr + 24, 19, 15, instr + 24, 11, 7) &&  // check the rs1 field and the rd field of slli
-           compare_instr_field(instr + 28, 19, 15, instr + 24, 11, 7) &&  // check the rs1 field of addi and the rd field of slli
-           compare_instr_field(instr + 28, 19, 15, instr + 28, 11, 7);    // check the rs1 field and the rd field of addi
+    const address lui = instr;
+    const Register lui_rd = extract_rd(lui);
+    const address addi1 = lui + instruction_size;
+    const Register addi1_rs1 = extract_rs1(addi1);
+    const Register addi1_rd  = extract_rd(addi1);
+    const address slli1 = addi1 + instruction_size;
+    const Register slli1_rs1 = extract_rs1(slli1);
+    const Register slli1_rd  = extract_rd(slli1);
+    const address addi2 = slli1 + instruction_size;
+    const Register addi2_rs1 = extract_rs1(addi2);
+    const Register addi2_rd  = extract_rd(addi2);
+    const address slli2 = addi2 + instruction_size;
+    const Register slli2_rs1 = extract_rs1(slli2);
+    const Register slli2_rd  = extract_rd(slli2);
+    const address addi3 = slli2 + instruction_size;
+    const Register addi3_rs1 = extract_rs1(addi3);
+    const Register addi3_rd  = extract_rd(addi3);
+    const address slli3 = addi3 + instruction_size;
+    const Register slli3_rs1 = extract_rs1(slli3);
+    const Register slli3_rd  = extract_rd(slli3);
+    const address addi4 = slli3 + instruction_size;
+    const Register addi4_rs1 = extract_rs1(addi4);
+    const Register addi4_rd  = extract_rd(addi4);
```

Comment:
I don't think so many local variables are a good idea, at least it doesn't do much to improve readability, except to make the code verbose.

Reply:
In fact - I have the same thought though after revising the code. Seems sadly I cannot find a way to extract them only once and in the meantime keep the readability without introducing new local variables. I don't know - maybe turning back to [my original version](https://github.com/openjdk/riscv-port/blob/92273c63986eef7126194f32c28a915fb0abf11c/src/hotspot/cpu/riscv/nativeInst_riscv.hpp#L100-L116) is a choice?

Reply:
Yes. In my personal opinion, it would be better.

Reply:
Thanks for the decision - changed back `check_*_data_dependency` to the first version, and tested a `test/jtreg/hotspot/compiler` on Qemu.

Reply:
I think I can live with this assuming the host compiler will help us eliminate redundancy here.